app/palette/csrf_storage.py

```diff
@@ -0,0 +1,155 @@
+import logging
+import uuid
+import threading
+import numbers
+import datetime
+
+
+def generate_token():
+    """ Generates a new token """
+    return str(uuid.uuid1())
+
+
+class LoggingComponent():
+    """ Generic class with an injected logger and a log() method """
+    def __init__(self, logger):
+        self._logger = logger
+
+    def log(self, lvl, msg, *args, **kwargs):
+        """ Attempts to call the logger (if provided) by forwarding all args """
+        if self._logger:
+            self._logger.log(lvl, msg, *args, **kwargs)
+
+
+
+class InMemoryCsrfStorage(LoggingComponent):
+    """ An implementation of an in-memory dictionary used for CSRF storage """
+    def __init__(self,
+                 logger=logging.getLogger('in_memory_csrf_storage')):
+        LoggingComponent.__init__(self, logger)
+        # Our token storage
+        self._tokens = {}
+        # Create a lock to protect the tokens dict
+        self._lock = threading.Lock()
+
+    def create_token(self, old_token, current_time):
+        token = generate_token()
+
+        # Ensure locking
+        with self._lock:
+            # remove old token
+            if old_token is not None and old_token in self._tokens:
+                self.log(logging.DEBUG, "deleting old token: %s", old_token)
+                del self._tokens[old_token]
+
+            # update new token
+            self.log(logging.DEBUG, "saving new token: %s", token)
+            self._tokens[token] = current_time
+            return token
+
+    def validate_token(self, token, current_time):
+        self.log(logging.DEBUG, "validating token: %s", token)
+        with self._lock:
+            return token in self._tokens
+
+
+
+def format_sql(str, *args):
+    """ Format an SQL string and escape things """
+    def escape(val):
+        if isinstance(val, basestring):
+            return "%s" % val
+        elif isinstance(val, numbers.Number):
+            return str(val)
```

Comment:
Is `str` really the type in this line or the argument of `format_sql`?

Reply:
should be fixed in commit 0b317b3a1d5d2b112f4a0599a7e9df882c0e9701